<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started â€¢ IBRAP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started">
<meta property="og:description" content="IBRAP">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IBRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installing_ibrap.html">Installing &amp; Preparing IBRAP</a>
    </li>
    <li>
      <a href="../articles/using_IBRAP.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/data_transform.html">Data Transformation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_IBRAP_files/header-attrs-2.10.1/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started</h1>
                        <h4 data-toc-skip class="author">Connor H Knight</h4>
            
            <h4 data-toc-skip class="date">11/08/2021</h4>
      
      
      <div class="hidden name"><code>using_IBRAP.Rmd</code></div>

    </div>

    
    
<div id="analysing-2994-bone-marrow-mononuclear-cells-bmmc-from-a-single-sample-" class="section level1">
<h1 class="hasAnchor">
<a href="#analysing-2994-bone-marrow-mononuclear-cells-bmmc-from-a-single-sample-" class="anchor" aria-hidden="true"></a>Analysing 2994 Bone Marrow Mononuclear Cells (BMMC) from a single sample.</h1>
<p>To demonstrate IBRAP we enlisted a sample produced by <a href="https://insight.jci.org/articles/view/124928" class="external-link">Oetjen</a> <em>et al</em>. This can be downloaded using the following link:</p>
<p><a href="https://www.dropbox.com/sh/f2hhztiyzo04yt9/AACeK5nZBcQ7Fnz02hgF6GG5a?dl=0" class="external-link uri">https://www.dropbox.com/sh/f2hhztiyzo04yt9/AACeK5nZBcQ7Fnz02hgF6GG5a?dl=0</a></p>
<div id="initiating-the-ibrap-s4-object" class="section level2">
<h2 class="hasAnchor">
<a href="#initiating-the-ibrap-s4-object" class="anchor" aria-hidden="true"></a>Initiating the IBRAP S4 object</h2>
<p>Since our datasets are produced by 10x genomics and aligned using CellRanger, 3 subsequent files are produced to integrate these files into a count matrix.</p>
<pre><code>
bmmc &lt;- Read10X_output(directory = '~Downloads/marrow_A', 
                       matrix.file = 'matrix.mtx', 
                       genes.file = 'genes.tsv', 
                       barcodes.file = 'barcodes.tsv')
                       </code></pre>
<p>Droplet-based technologies are normally accompanied by specific conundra that requires omission, namely: multiplets and ambient RNA. With IBRAP you can apply a python package called scrublet (perform.scrublet()) and decontX (perform.decontX()) from the celda package in R for removal, respectively. We supply our count matrix to the functions. There is no defined order of applciation, but logically we assume scrublet removal followed by decontX.</p>
<pre><code>
# We simply supply our generated count matrix to both functions
bmmc &lt;- perform.scrublet(counts = bmmc)

bmmc &lt;- perform.decontX(counts = bmmc)
</code></pre>
<p>Now we can initiate our IBRAP class object ready for downstream analyses. Refer to methods-object and IBRAP-object for further detail.</p>
<pre><code>
# whilst creating the object we begin filtering the cells and genes using min.cells and min.features
bmmc &lt;- createIBRAPobject(counts = bmmc,
                          original.project = 'bmmc',
                          method.name = 'RAW',
                          min.cells = 3,
                          min.features = 200)
                          </code></pre>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor" aria-hidden="true"></a>Quality Control</h2>
<p>Current best practices for quality assuring scRNA-seq datasets are outlined in numerous <a href="https://www.embopress.org/doi/full/10.15252/msb.20188746" class="external-link">publications</a>. Including:</p>
<table class="table">
<colgroup>
<col width="13%">
<col width="86%">
</colgroup>
<thead><tr class="header">
<th align="left">QC.Metric</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Mitochondrial Fraction</td>
<td align="left">An abnormally high mitochondrail fraction indicates improper wet lab preparation</td>
</tr>
<tr class="even">
<td align="left">Total RNA</td>
<td align="left">Low RNA quanitity indicates droplets containing no cells whereas, disproportionately high RNA indicates a droplet containing more than one cell</td>
</tr>
<tr class="odd">
<td align="left">Total Genes</td>
<td align="left">likewise, a low or high number of genes typically correspond with total RNA</td>
</tr>
</tbody>
</table>
<p>The total RNA and Total genes are calculated upon the creation of the IBRAP object. However, percentage mitochondrial genes requires a second function since mitochondrial genes have a varying prefix in datasets. Thus, we use find_percentage_genes():</p>
<pre><code>
bmmc &lt;- find_percentage_genes(object = celseq2, pattern = '^MT-',
                              assay = 'RAW', 
                              slot = 'counts',
                              column.name = 'RAW_percent.mt')
                              </code></pre>
<p>We can visualise this information in two ways: using violin plots (plot.qc.vln()) or comparing two metrices with scatter plots (plot.qc.scatter()). The latter enables us to observe correlations between the metrices.</p>
<pre><code>
plot.QC.vln(object = celseq2, 
            metadata.columns = c('RAW_total.features', 
                                 'RAW_total.counts', 
                                 'RAW_percent.mt'))
                                 </code></pre>
<p>Filtration is subsequently applied to remove problematic cells.</p>
<pre><code>
bmmc &lt;- filter_IBRAP(RAW_total.features &lt; 2500 &amp; RAW_total.counts &gt; 200 &amp; RAW_percent.mt &lt; 8)
</code></pre>
</div>
<div id="data-transformation" class="section level2">
<h2 class="hasAnchor">
<a href="#data-transformation" class="anchor" aria-hidden="true"></a>Data Transformation</h2>
<p>To accommodate downstream analyses, we must promote a normal distribution and numerical scaling in our data.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Connor Knight.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
